prov/mrail/src/mrail_fabric.c

```diff
@@ -73,16 +74,36 @@ static struct fi_ops_fabric mrail_fabric_ops = {
 int mrail_fabric_open(struct fi_fabric_attr *attr, struct fid_fabric **fabric,
 		      void *context)
 {
-	struct fi_info *fi;
+	struct fi_info *fi, *info, *hints;
 	struct mrail_fabric *mrail_fabric;
 	size_t i;
 	int ret;
 
 	mrail_fabric = calloc(1, sizeof(*mrail_fabric));
 	if (!mrail_fabric)
 		return -FI_ENOMEM;
+	ret = mrail_get_info_cached(attr->name, attr->prov_name, &info);
 
-	mrail_fabric->info = mrail_get_info_cached(attr->name);
+	if (ret) {
+		FI_WARN(&mrail_prov,
+			FI_LOG_CORE,
+			"No cached data found, running getinfo with prov name\n");
+		hints = fi_allocinfo();
+		if (!hints)
+			return -FI_ENOMEM;
+		hints->fabric_attr->prov_name = attr->prov_name;
+		ret = fi_getinfo(attr->api_version, NULL, NULL, 0, hints, &fi);
+		fi_freeinfo(hints);
+		if (ret)
+			return ret;
+		ret = mrail_get_info_cached(fi->fabric_attr->name,
+					    fi->fabric_attr->prov_name,
+					    &info);
+		if (ret)
+			return ret;
+	}
```

Comment:
Please move this if statement code into a separate function to improve readability.